### DISCUSSION TITLE
Setting @context for JSON-LD parsing ?

### DISCUSSION THREAD

**Discussion by tfrancart** [eclipse-rdf4j/rdf4j#4059] [Q&A] [answered]
Hello !

I would like to parse JSON files that either do not declare @context at all, and pass in the @context programmatically when parsing. 
I don't see how I can do that.
I can use directly the jsonld-java underlying implementation, but I would like to get as an output an RDFJ4 repository, but JSONLDInternalTripleCallback, as the name suggests... is private.

How can I specify the @context programmatically when parsing ?
Similarly, if the JSON points to a @context to a certain fake URL, how can I override that @context with mine, at parsing time ?

I found this : https://github.com/eclipse/rdf4j/issues/395, is this related ?

Little code snippet would be much appreciated !

Thomas

**Comment by ate47** [comment, marked as answer]
It seems the library can't do that, the closest I was able to find is if the the `"@context"` URL is a fake one, but still an URL, you can use the `JSONLDSettings.DOCUMENT_LOADER` option to override the context download

I wrote this little example to read the `original.json` file and convert it into a ttl file.

**original.json** (the [Wikipedia example](https://en.wikipedia.org/wiki/JSON-LD))

```json
{
  "@context": "http://bad.com/url",
  "@id": "https://me.example.com",
  "@type": "Person",
  "name": "John Smith",
  "homepage": "https://www.example.com/"
}
```

the removed context is
```json
"@context": {
    "name": "http://xmlns.com/foaf/0.1/name",
    "homepage": {
        "@id": "http://xmlns.com/foaf/0.1/workplaceHomepage",
        "@type": "@id"
    },
    "Person": "http://xmlns.com/foaf/0.1/Person"
 },
```

Parser code
```java
try (InputStream inputStream = new FileInputStream("original.json");
        OutputStream stream = new FileOutputStream("original.ttl")) {
    RDFParser rdfParser = Rio.createParser(RDFFormat.JSONLD);

    // replace the document loader used to read remote documents
    rdfParser.set(JSONLDSettings.DOCUMENT_LOADER, new DocumentLoader() {

        @Override
        public RemoteDocument loadDocument(String url) throws JsonLdError {
            // here you can check the URL if you want

            // this method return a RemoteDocument, the first argument isn't important for the
            // context, but the second argument should be a Map<String, Object> describing your
            // json tree
            return new RemoteDocument("whatever", Map.<String, Object>of(
                    "@context", Map.<String, Object>of(
                            "name", "http://xmlns.com/foaf/0.1/name",
                            "homepage", Map.<String, Object>of(
                                "@id", "http://xmlns.com/foaf/0.1/workplaceHomepage",
                                "@type", "@id"
                            ),
                            "Person", "http://xmlns.com/foaf/0.1/Person"
                    )
            ));
        }
    });

    // here I'm using a RDFWriter to handle it, but you can pass it to a repository
    RDFWriter writer = Rio.createWriter(RDFFormat.TURTLE, stream);
    rdfParser.setRDFHandler(writer);

    // parse the document
    rdfParser.parse(inputStream);
}
```

Result:

**original.ttl**
```turtle

<https://me.example.com> a <http://xmlns.com/foaf/0.1/Person>;
  <http://xmlns.com/foaf/0.1/name> "John Smith";
  <http://xmlns.com/foaf/0.1/workplaceHomepage> <https://www.example.com/> .
```

This solution is a bit hacky I admit

**Reply by tfrancart** [reply]
Thanks ! I think this is how far we can get, I guess.
It surprises me that the context cannot be provided completely programmaticaly (that is, even without the JSON file declaring a @context), because the context can be provided on top of normal JSON documents in a Link : header : https://www.w3.org/TR/json-ld11/#interpreting-json-as-json-ld, so theoretically a plain JSON document can benefit from a separate, non-declared, JSON-LD context.
